### PR TITLE
Don't allocate in BeginChunkBytes

### PR DESCRIPTION
**Pull request by benaadams**
Use the `BufferWriter`'s Span instead

Think allocations are up by factor of 1024 in the benchmark (i.e. x IterationCount for the pre)
```diff
                Method | DataLength |      Mean |          Op/s |  Gen 0 | Allocated |
 --------------------- |----------- |----------:|--------------:|-------:|----------:|
- WriteBeginChunkBytes |          0 |  33.30 ns |  30,030,302.5 | 0.7324 |     40 KB |
+ WriteBeginChunkBytes |          0 |   8.09 ns | 123,639,689.1 | 0.7324 |       0 B |
- WriteBeginChunkBytes |          1 |  34.48 ns |  28,998,176.5 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |          1 |   8.40 ns | 119,093,216.2 | 0.6714 |       0 B |
- WriteBeginChunkBytes |         16 |  34.07 ns |  29,348,584.0 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |         16 |   9.01 ns | 110,966,446.7 | 0.6714 |       0 B |
- WriteBeginChunkBytes |        256 |  45.12 ns |  22,164,899.8 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |        256 |  10.81 ns |  92,489,663.4 | 0.6714 |       0 B |
- WriteBeginChunkBytes |       4096 |  33.05 ns |  30,257,577.0 | 0.7324 |     40 KB |
+ WriteBeginChunkBytes |       4096 |  11.54 ns |  86,654,067.4 | 0.7324 |       0 B |
- WriteBeginChunkBytes |      65536 |  45.85 ns |  21,809,105.6 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |      65536 |  12.69 ns |  78,783,888.6 | 0.6714 |       0 B |
- WriteBeginChunkBytes |    1048576 |  33.94 ns |  29,459,649.1 | 0.7324 |     40 KB |
+ WriteBeginChunkBytes |    1048576 |  13.92 ns |  71,856,593.8 | 0.7324 |       0 B |
- WriteBeginChunkBytes |   16777216 |  33.30 ns |  30,028,526.6 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |   16777216 |  15.35 ns |  65,169,537.2 | 0.6714 |       0 B |
```

/cc @pakrym @davidfowl

### CODE REVIEW

**Comment by benaadams** [comment]
Not sure how to elide bounds checks here (and I assume `Unsafe.Add` is going a bit fast and loose at this level of the stack?)

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs [hunk 1]
```diff
@@ -19,40 +19,40 @@ private static ArraySegment<byte> CreateAsciiByteArraySegment(string text)
...
+            total |= (count > 0x000f) ? 0x04 : 0x00;
 
-            var offset = 7 - (total >> 2);
-            return new ArraySegment<byte>(bytes, offset, 10 - offset);
+            count = (total >> 2) + 3;
+
+            var offset = 0;
+            for (shift = total; shift >= 0; shift -= 4)
+            {
+                span[offset] = _hex[((dataCount >> shift) & 0x0f)];
```
Comment (Tornhoof) [inline src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Would converting _hex to a ROS help here?

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs [hunk 2]
```diff
@@ -19,40 +19,40 @@ private static ArraySegment<byte> CreateAsciiByteArraySegment(string text)
...
-            start.Write(new ReadOnlySpan<byte>(chunkSegment.Array, chunkSegment.Offset, chunkSegment.Count));
-            return chunkSegment.Count;
+            // 10 bytes is max length + \r\n
+            start.Ensure(10);
+
+            var count = BeginChunkBytes(dataCount, start.Span);
+            start.Advance(count);
         }
 
         internal static void WriteEndChunkBytes(ref BufferWriter<PipeWriter> start)
```
Comment (Tornhoof) [inline src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Not really Part of this PR, as it is WriteEnd, but ist there are reason why _endChunkBytes isn't a ROS, but an ArraySegment and then a ROS is created for each write? Or Just "old" code, prior to all those ROS optimizations.

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Old code

**Comment by davidfowl** [comment]
@benaadams perf tests?

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs (hunk 2 above)
Comment (Tornhoof) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: I was specifically referring to that static ROS<byte> property trick, which is currently en vogue everywhere: https://github.com/dotnet/coreclr/pull/20768

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Sure, why not :)

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: And since we are modernising can change them into `this ref` extension methods

**Comment by benaadams** [comment]
Don't think BDN is dividing allocations by iterations, howevers:
```diff
                Method | DataLength |      Mean |          Op/s |  Gen 0 | Allocated |
 --------------------- |----------- |----------:|--------------:|-------:|----------:|
- WriteBeginChunkBytes |          0 |  33.30 ns |  30,030,302.5 | 0.7324 |     40 KB |
+ WriteBeginChunkBytes |          0 |   8.09 ns | 123,639,689.1 | 0.7324 |       0 B |
- WriteBeginChunkBytes |          1 |  34.48 ns |  28,998,176.5 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |          1 |   8.40 ns | 119,093,216.2 | 0.6714 |       0 B |
- WriteBeginChunkBytes |         16 |  34.07 ns |  29,348,584.0 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |         16 |   9.01 ns | 110,966,446.7 | 0.6714 |       0 B |
- WriteBeginChunkBytes |        256 |  45.12 ns |  22,164,899.8 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |        256 |  10.81 ns |  92,489,663.4 | 0.6714 |       0 B |
- WriteBeginChunkBytes |       4096 |  33.05 ns |  30,257,577.0 | 0.7324 |     40 KB |
+ WriteBeginChunkBytes |       4096 |  11.54 ns |  86,654,067.4 | 0.7324 |       0 B |
- WriteBeginChunkBytes |      65536 |  45.85 ns |  21,809,105.6 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |      65536 |  12.69 ns |  78,783,888.6 | 0.6714 |       0 B |
- WriteBeginChunkBytes |    1048576 |  33.94 ns |  29,459,649.1 | 0.7324 |     40 KB |
+ WriteBeginChunkBytes |    1048576 |  13.92 ns |  71,856,593.8 | 0.7324 |       0 B |
- WriteBeginChunkBytes |   16777216 |  33.30 ns |  30,028,526.6 | 0.6714 |     40 KB |
+ WriteBeginChunkBytes |   16777216 |  15.35 ns |  65,169,537.2 | 0.6714 |       0 B |
```

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs [hunk 3]
```diff
@@ -5,59 +5,58 @@
...
+
+            var count = BeginChunkBytes(dataCount, start.Span);
+            start.Advance(count);
         }
 
-        internal static void WriteEndChunkBytes(ref BufferWriter<PipeWriter> start)
+        internal static void WriteEndChunkBytes(this ref BufferWriter<PipeWriter> start)
         {
-            start.Write(new ReadOnlySpan<byte>(_endChunkBytes.Array, _endChunkBytes.Offset, _endChunkBytes.Count));
+            start.Write(EndChunkBytes);
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Done

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs (hunk 1 above)
Comment (Tornhoof) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: You can do the same optimization as for _endChunkBytes  for _hex too, use the static property and create the hex bytes manually.

**Review by pakrym — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs (hunk 1 above)
Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Not sure that would have any benefit?

Comment (pakrym) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Let's keep this PR limited to removing the allocation.

**Review by davidfowl — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs [hunk 4]
```diff
@@ -5,59 +5,62 @@
...
 using System.IO.Pipelines;
 using System.Text;
+using System.Runtime.CompilerServices;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
 {
     internal static class ChunkWriter
     {
-        private static readonly ArraySegment<byte> _endChunkBytes = CreateAsciiByteArraySegment("\r\n");
         private static readonly byte[] _hex = Encoding.ASCII.GetBytes("0123456789abcdef");
```
Comment (pentp) [inline src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: This could be changed to a `ReadonlySpan<byte>` returning property (or inlined in the method) - it would be compiled as just static data and `startHex` will be a constant then. This would get rid of the static .cctor also.

**Comment by benaadams** [comment]
Some failures in the main ci; checking locally if releated

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs [hunk 5]
```diff
@@ -5,59 +5,62 @@
...
-        internal static void WriteEndChunkBytes(ref BufferWriter<PipeWriter> start)
+        internal static void WriteEndChunkBytes(this ref BufferWriter<PipeWriter> start)
         {
-            start.Write(new ReadOnlySpan<byte>(_endChunkBytes.Array, _endChunkBytes.Offset, _endChunkBytes.Count));
+            start.Ensure(2);
+            var span = start.Span;
+
+            // CRLF done in reverse order so the 1st index will elide the bounds check for the 0th index
+            span[1] = (byte)'\n';
+            span[0] = (byte)'\r';
```
Comment (benaadams) [inline src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Needs `start.Advance(2);`

File: src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/ChunkWriter.cs]: Missed and then added https://github.com/aspnet/AspNetCore/pull/5688/commits/5c0181dfabccdd518c2aabc68463d61d2b5783fd the Advance for this

**Comment by benaadams** [comment]
2 unrelated errors
```
Microsoft.AspNetCore.FunctionalTests.Microsoft.AspNetCore.Tests.WebHostFunctionalTests.RunsInIISExpressInProcess
Assert.Equal() Failure
          ↓ (pos 0)
Expected: CreateDefaultBuilderApp
Actual:   <!DOCTYPE html>\r\n<html lang="en-US" ns···
          ↑ (pos 0)
   at Microsoft.AspNetCore.Tests.WebHostFunctionalTests.RunsInIISExpressInProcess()
   in /_/src/DefaultBuilder/test/Microsoft.AspNetCore.FunctionalTests/WebHostFunctionalTests.cs:line 179
--- End of stack trace from previous location where exception was thrown ---
```
```
ServerComparison.FunctionalTests.ResponseTests.ResponseFormats_Chunked(variant: 
 Server: IISExpress, TFM: netcoreapp3.0, Type: Portable, Arch: x64, ANCM: V2, Host: InProcess)
/chunked, chunked?
Expected: True
Actual:   (null)
   at ServerComparison.FunctionalTests.ResponseTests.CheckChunkedAsync(HttpClient client, ILogger logger)
    in /_/src/Servers/test/FunctionalTests/ResponseTests.cs:line 211
   at ServerComparison.FunctionalTests.ResponseTests.ResponseFormats(TestVariant variant, Func`3 scenario, String testName)
    in /_/src/Servers/test/FunctionalTests/ResponseTests.cs:line 125
--- End of stack trace from previous location where exception was thrown ---
```

**Comment by benaadams** [comment]
Opened issues for the errors #6135 and #6136

**Comment by pakrym** [comment]
While `RunsInIISExpressInProcess` constantly fails, I've never seen `ResponseFormats_Chunked` failing in other RP.  Doesn't seem to be connected at all but I'll restart it again.

**Comment by pakrym** [comment]
I lied, https://github.com/aspnet/AspNetCore/pull/6119/checks?check_run_id=44328480 has it too.
